### PR TITLE
Add Kuryr and Genie CNIs

### PR DESCRIPTION
**Pull request by yuvalif**
This PR allows the VM to get multiple networks/interfaces via Multus, Kuryr or Genie CNIs.

_**Currently ~~unit and~~ functional tests are missing.**_

**Release Notes:**
To add multiple networks/interfaces to a Virtual Machine Instance, sever options now exists:
* For Multus add the following network definitions:
```yaml
networks:
  - multus:
      networkName: net-a
    name: net-a
```
Where, ```networkName``` is the name of the ```NetworkAttachment``` CRD to be used. In case of Multus, the actual network is provided by another CNI (flannel, weave etc.) that must be deployed beforehand.
* For Genie add the following network definitions:
```yaml
networks:
  - genie:
      networkName: flannel
    name: net-a
```
Where, ```networkName``` is the name the actual CNI that would provide the network (flannel, weave etc.), and must be deployed beforehand.
* For Kuryr add the following network definitions:
```yaml
networks:
  - kuryr:
      networkName: net-a
    name: net-a
```
Where, ```networkName``` is the name of the ```NetworkAttachment``` CRD to be used. Kuryr does not delegate the network definition to another CNI, and will provide the networking based on an existing openstack networking solution which has to be deployed beforehand.

> Notes: 
> * Only one of the above CNIs (Multus, Genie, Kuryr) could be used across all networks
> * Pod network could be combined with Multus or Kuryr, where it will get ```eth0``` in the pod without any ```NetworkAttachment``` CRD. The network definition will be taken from the default of Multus or Kuryr as configuration under ```/etc/cni/net.d/``` 
> * Pod network cannot be combined with Genie CNI. However, if Genie CNI is deployed, and only Pod network is configured, it would be allowed, as the default CNI will be taken from the default of Genie as configuration under ```/etc/cni/net.d/```
> * In all cases, "name" is used as a reference in the "interface" definitions

```release-note
see release notes above
```

### CODE REVIEW

**Comment by yuvalif** [comment]
[spike: what is Genie vs Mutus? can we use it?](https://trello.com/c/rCh4snhV/156-spike-what-is-genie-vs-mutus-can-we-use-it)

**Comment by yuvalif** [comment]
[spike: what is Genie vs Mutus? can we use it?](https://trello.com/c/rCh4snhV/156-spike-what-is-genie-vs-mutus-can-we-use-it)

**Comment by yuvalif** [comment]
[spike: what is Kuryr? does it play well with CNV?](https://trello.com/c/7x8WTBuF/155-spike-what-is-kuryr-does-it-play-well-with-cnv)

File: pkg/virt-api/validating-webhook/validating-webhook.go [hunk 1]
```diff
@@ -608,29 +608,77 @@ func ValidateVirtualMachineInstanceSpec(field *k8sfield.Path, spec *v1.VirtualMa
...
+				if network.Cni.NetworkName == "" {
+					causes = append(causes, metav1.StatusCause{
+						Type:    metav1.CauseTypeFieldValueRequired,
+						Message: fmt.Sprintf("CNI network must have a networkName"),
+						Field:   field.Child("networks").Index(idx).Child("cni").String(),
+					})
+				}
+
+				if network.Cni.Multus != nil {
+					if kuryrExists || genieExists {
```
Comment (davidvossel) [inline pkg/virt-api/validating-webhook/validating-webhook.go]: this logic is difficult to maintain as more network devices are added. Take a look at how the volume source handles ensuring only a single item is set.

File: pkg/api/v1/schema.go [hunk 2]
```diff
@@ -783,8 +783,8 @@ type Network struct {
 // ---
 // +k8s:openapi-gen=true
 type NetworkSource struct {
-	Pod    *PodNetwork    `json:"pod,omitempty"`
-	Multus *MultusNetwork `json:"multus,omitempty"`
+	Pod *PodNetwork `json:"pod,omitempty"`
+	Cni *CniNetwork `json:"cni,omitempty"`
```
[the remark this answers is not in the document]
Comment (yuvalif) [reply pkg/api/v1/schema.go]: several reasons:
* they are all providing multiple networks (pod network is only one)
* they all require a "network name" attribute
* they are mutually exclusive from one another (and not from pod)
* there are quiet a bit of places in the code where there are checks related to them only

**Comment by yuvalif** [comment]
> Our test matrix is going to explode if we have to have a different provider for each network and k8s/os version. I remember this being brought up briefly during a meeting at one point. Is there a plan for how to handle functional testing?

@davidvossel for genie, we could probably reuse the multus provider. @SchSeba how difficult would that be?
For kuryr, I don't see how this could be automatic testing - openstack provider is probably out-of-the question. 

A different option would be to write a CNI simulator that would simulate the behavior for the different providers? Its not ideal, as we will have to make sure that the bugs are not there, but might be an acceptable compromise.

File: pkg/api/v1/schema.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (davidvossel) [reply pkg/api/v1/schema.go]: > they are all providing multiple networks (pod network is only one)

Every network source needs to represent a single network, whether it be a pod network, multus network, genie, etc... The difference is that some network sources (like multus) can have multiple entries in the "networks" list. 

There's a 1-1 relationship between NetworkSources and interfaces in the DomainSpec. 

> they all require a "network name" attribute

you can reuse the same struct internally if you want and still keep a flat structure.

```
type NetworkSource struct {
    Pod *PodNetwork `json:"pod,omitempty"
    Multus *CniNetwork `json:"multus,omitempty"`
    Genie *CniNetwork `json:"genie,omitempty"`
    Kuryr *CniNetwork `json:"kuryr,omitempty"`
}
```

> they are mutually exclusive from one another (and not from pod)

All the network sources should be mutually exclusive from one another within the NetworkSource struct. Only a single struct can be set per a NetworkSource. This should be modeled after how the VolumeSource works.

> there are quiet a bit of places in the code where there are checks related to them only

This seems like an internal implementation detail. This internal coupling shouldn't be something that's reflected in the API unless it somehow makes sense from a user perspective.

[the remark this answers is not in the document]
Comment (yuvalif) [reply pkg/api/v1/schema.go]: > All the network sources should be mutually exclusive from one another within the NetworkSource struct. Only a single struct can be set per a NetworkSource. This should be modeled after how the VolumeSource works.

not sure that it matters for this discussion, but they are mutually exclusive at VMI level, not only at NetworkSource level.

> This seems like an internal implementation detail. This internal coupling shouldn't be something that's reflected in the API unless it somehow makes sense from a user perspective.

To me, both options looks OK. I preferred the one where the code is cleaner, but I'm also good with less hierarchies in the API.

**Comment by fabiand** [comment]
WRT functional tests.

multus is the interestign bit right now.

But with more and more of integrations like this I wonder how we can do this.

We should only call it supported if we have real e2e tests for it.
A compromise would be to have the test, but not run them always, instead focus on calling the expected APi correctly (aka set the correct labels/annotations).

Maybe a regular time based job to test those integrations.

**Comment by fabiand** [comment]
Ping?

**Comment by yuvalif** [comment]
Please review: https://github.com/kubevirt/kubevirt/pull/1590 instead (Kuryr was removed from the new PR).

**Pull request closed without merging**